common/OpTestSSH.py

```diff
@@ -0,0 +1,180 @@
+from OpTestConstants import OpTestConstants as BMC_CONST
+import OpTestSystem
+try:
+    from common import OPexpect
+except ImportError:
+    import OPexpect
+
+class ConsoleState():
+    DISCONNECTED = 0
+    CONNECTED = 1
+
+def set_system_to_UNKNOWN(system):
+    s = system.get_state()
+    system.set_state(OpTestSystem.OpSystemState.UNKNOWN)
+    return s
+
+class OpTestSSH():
+    def __init__(self, host, username, password, logfile=sys.stdout, port=22,
+            prompt=None):
+        self.state = ConsoleState.DISCONNECTED
+        self.host = host
+        self.username = username
+        self.password = password
+        self.port = port
+        self.logfile = logfile
+        self.prompt = prompt
+        self.system = None
+
+    def set_system(self, system):
+        self.system = system
+
+    def terminate(self):
+        if self.state == ConsoleState.CONNECTED:
+            self.console.terminate()
+            self.state = ConsoleState.DISCONNECTED
+
+    def close(self):
+        if self.state == ConsoleState.DISCONNECTED:
+            return
+        try:
+            self.console.send("\r")
+            self.console.send('~.')
+            self.console.expect(pexpect.EOF)
+            self.console.close()
+        except pexpect.ExceptionPexpect:
+            raise "SSH Console: failed to close ssh console"
+        self.console.terminate()
+        self.state = ConsoleState.DISCONNECTED
+
+    def connect(self):
+        if self.state == ConsoleState.CONNECTED:
+            self.console.terminate()
+            self.state = ConsoleState.DISCONNECTED
+
+        cmd = ("sshpass -p %s " % (self.password)
+               + " ssh -q"
+               + " -p %s" % str(self.port)
+               + " -l %s %s" % (self.username, self.host)
+               + " -o'RSAAuthentication=no' -o 'PubkeyAuthentication=no'"
+               + " -o 'UserKnownHostsFile=/dev/null' "
+               + " -o 'StrictHostKeyChecking=no'"
+           )
+
+        print cmd
+        consoleChild = OPexpect.spawn(cmd,logfile=self.logfile,
+                failure_callback=set_system_to_UNKNOWN,
+                failure_callback_data=self.system)
+        self.state = ConsoleState.CONNECTED
+        self.console = consoleChild
+        # Users expecting "Host IPMI" will reference console.sol so make it available
+        self.sol = self.console
+        return consoleChild
+
+    def get_console(self):
+        if self.state == ConsoleState.DISCONNECTED:
+            self.connect()
+
+        count = 0
+        while (not self.console.isalive()):
+            print '# Reconnecting'
+            if (count > 0):
+                time.sleep(1)
+            self.connect()
+            count += 1
+            if count > 120:
+                raise "SSH: not able to get console"
+
+        return self.console
+
+    def run_command(self, command, timeout=60):
+        console = self.get_console()
+
+        if self.prompt:
+            prompt = self.prompt
```

Comment:
Can we set the prompt of BMC busy box to `[PEXPECT]\#` . Please use the previous code which sets the unique prompt for busy box as well. As some of the BMC command output as `#` in it which run_command interprets it as prompt, so we will end up in wrong output collection.
So please re-use this existing code to set the unique prompt.
```
        p.login(self.cv_bmcIP, self.cv_bmcUser, self.cv_bmcPasswd, auto_prompt_reset=False)
        p.sendline()
        p.prompt(timeout=60)
        p.sendcontrol('l')
        p.expect(r'.+#')
        p.sendline('PS1=[PEXPECT]\#')
        p.expect("\n") # from us, because echo
        l_rc = p.expect("\[PEXPECT\]#$")
        if l_rc == 0:
            print "Shell prompt changed"
        else:
            raise Exception("Failed during change of shell prompt")
```
For console we will keep the `\[console-pexpect\]#$"` as unique prompt.

Reply:
Right, just `#` was set for some previously so I copied it, but testing it behaves a lot nicer with `[PEXPECT]#`. Is there any reason not to use `[console-pexpect]` everywhere though?

Reply:
When running tests we should be able to know which command executes where we kept that distinguish, for IPMI SOL console and OpenBMC ssh console(port 2200)(console based) we kept as `[console-pexpect]#` and for all other Host SSH and BMC SSH sessions(SSH based) we are using `[PEXPECT]#` .

Reply:
Cool, I'll make that consistent then. For easy readability would it make more sense to use `[host-console]`, `[host-ssh]`, and `[bmc-ssh]` then, rather than just `[PEXPECT]`?

Reply:
Yes, That makes sense. Please use them.